src/wmtk/PolyMesh.hpp

```diff
@@ -0,0 +1,264 @@
+    Eigen::Ref<const VectorXl> opp);
+
+/**
+ * Build orbits following next id recorded in perm.
+ *
+ * @param perm: a permutation given by a list of non-repeating integers in the range 0..len(perm)
+ * @return cycles: a list of lists, each list represents a cycle of perm
+ */
+std::vector<std::vector<int>> build_orbits(const std::vector<int>& perm);
+
+/**
+ * End of Helper Functions
+ */
+
+namespace wmtk {
+
+
+class PolyMesh : public Mesh
+{
+public:
+    PolyMesh();
+    PolyMesh(const PolyMesh& o);
+    PolyMesh(PolyMesh&& o);
+    PolyMesh& operator=(const PolyMesh& o);
+    PolyMesh& operator=(PolyMesh&& o);
+
+    // TODO This only makes sense for simplices as currently named
+    // It is also potentially ambiguous as our top type is a face, but our basic mesh primitive
+    // is a halfedge
+    PrimitiveType top_simplex_type() const override { return PrimitiveType::Face; }
+
+    /**
+     * @brief Primitive for changing connectivity without changing the number of edges.
+     * 
+     * Requires t and s to have consistent orientation
+     */
+    Tuple splice(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Primitive for adding a bubble component with a single edge.
+     */
+    Tuple make_edge();
+
+    /**
+     * @brief Primitive for removing a bubble component with a single edge.
+     */
+    Tuple delete_bubble(const Tuple& t);
+
+    /**
+     * @brief Split the common face of t and s with the new edge between their respective
+     * vertices (with ccw orientation)
+     * 
+     * Requires t and s to share a face but not a vertex and have consistent orientation
+     */
+    Tuple split_face(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Join the two faces across the edge of t
+     * 
+     * Requires the edge of t to be adjacent to two distinct faces and have consistent orientation
+     */
+    Tuple join_face(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Split the common vertex of t and s with the new edge between the merge of their
+     * respective faces
+     * 
+     * Requires t and s to share a vertex but not a face and have consistent orientation
+     */
+    Tuple split_vertex(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Join the two endpoints the edge of t
+     * 
+     * Requires at least one endpoint of h to have valence > 1, and if vertices are distinct cannot
+     * be both boundary, unless the edge is boundary,
+     */
+    Tuple join_vertex(const Tuple& t);
+
+    /**
+     * @brief Mark a face as a hole in the surface.
+     */
+    Tuple make_hole(const Tuple& t);
+
+    /**
+     * @brief Mark a face as not a hole in the surface.
+     */
+    Tuple fill_hole(const Tuple& t);
+
+    // TODO These only make sense for triangle meshes, but we do want them in the interface
+    // since they can be implemented using splice and bubble operations
+    Tuple split_edge(const Tuple& t, Accessor<long>& hash_accessor) override;
```

Comment:
these will be implemented as an Operation struct that holds several other Operation structs internally.